applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c

```diff
@@ -306,17 +319,146 @@ static int channel_index_get(const struct bt_conn *conn, uint8_t *index)
 	return -EINVAL;
 }
 
-static bool parse_cb(struct bt_data *data, void *user_data)
+static void supported_sample_rates_print(uint16_t supported_sample_rates, enum bt_audio_dir dir)
+{
+	char supported_str[20] = "";
+
+	if (supported_sample_rates & BT_AUDIO_CODEC_LC3_FREQ_48KHZ) {
+		strcat(supported_str, "48, ");
+	}
+
+	if (supported_sample_rates & BT_AUDIO_CODEC_LC3_FREQ_24KHZ) {
+		strcat(supported_str, "24, ");
+	}
+
+	if (supported_sample_rates & BT_AUDIO_CODEC_LC3_FREQ_16KHZ) {
+		strcat(supported_str, "16, ");
+	}
+
+	if (dir == BT_AUDIO_DIR_SINK) {
+		LOG_DBG("Headset supports: %s kHz in sink direction", supported_str);
+	} else if (dir == BT_AUDIO_DIR_SOURCE) {
+		LOG_DBG("Headset supports: %s kHz in source direction", supported_str);
+	}
+}
+
+static bool sink_parse_cb(struct bt_data *data, void *user_data)
 {
 	if (data->type == BT_AUDIO_CODEC_LC3_FREQ) {
 		uint16_t temp = sys_get_le16(data->data);
 
-		if (temp & BT_AUDIO_CODEC_CAPABILIY_FREQ) {
+		supported_sample_rates_print(temp, BT_AUDIO_DIR_SINK);
+
+		/* Try with the preferred sample rate first */
+		switch (CONFIG_BT_AUDIO_PREF_SAMPLE_RATE_VALUE) {
+		case BT_AUDIO_CODEC_CONFIG_LC3_FREQ_48KHZ:
+			if (temp & BT_AUDIO_CODEC_LC3_FREQ_48KHZ) {
+				lc3_preset_sink = lc3_preset_sink_48_4_1;
+				*(bool *)user_data = true;
+				/* Found what we were looking for, stop parsing LTV */
+				return false;
+			}
+
+			break;
+
+		case BT_AUDIO_CODEC_CONFIG_LC3_FREQ_24KHZ:
+			if (temp & BT_AUDIO_CODEC_LC3_FREQ_24KHZ) {
+				lc3_preset_sink = lc3_preset_sink_24_2_1;
+				*(bool *)user_data = true;
+				/* Found what we were looking for, stop parsing LTV */
+				return false;
+			}
+
+			break;
+
+		case BT_AUDIO_CODEC_CONFIG_LC3_FREQ_16KHZ:
+			if (temp & BT_AUDIO_CODEC_LC3_FREQ_16KHZ) {
+				lc3_preset_sink = lc3_preset_sink_16_2_1;
+				*(bool *)user_data = true;
+				/* Found what we were looking for, stop parsing LTV */
+				return false;
+			}
+
+			break;
+		}
+
+		/* If no match with the preferred, revert to trying highest first */
+		if (temp & BT_AUDIO_CODEC_LC3_FREQ_48KHZ) {
+			lc3_preset_sink = lc3_preset_sink_48_4_1;
+			*(bool *)user_data = true;
+		} else if (temp & BT_AUDIO_CODEC_LC3_FREQ_24KHZ) {
+			lc3_preset_sink = lc3_preset_sink_24_2_1;
+			*(bool *)user_data = true;
+		} else if (temp & BT_AUDIO_CODEC_LC3_FREQ_16KHZ) {
+			lc3_preset_sink = lc3_preset_sink_16_2_1;
 			*(bool *)user_data = true;
 		}
+
+		/* Found what we were looking for, stop parsing LTV */
 		return false;
 	}
 
+	/* Did not find what we were looking for, continue parsing LTV */
+	return true;
+}
+
+static bool source_parse_cb(struct bt_data *data, void *user_data)
```

Comment:
Quite a bit of duplicated code here. Can we refactor into a static function?

Reply:
You mean a static function for each of the cases? I'm not able to see a way to do that as pretty much each line uses it's own variable, but I'm happy to discuss!

(I did rename `temp` to `lc3_freq_bit`, I think that is a bit more descriptive)